Use thread workers to update facility cases

kaapstorm opened the pull request:
## Summary

This is custom code for the ONSE ISS project. They are importing facility-level aggregated indicators from DHIS2 and saving them to facility cases in CommCare.

There are a few hundred cases involved. The import takes several hours. This change reduces that time to less than half an hour. The amount of time is not relevant when run in the background queue, but it does impact testing and running the import manually for the first import.

Quoting from [commit message](https://github.com/dimagi/commcare-hq/commit/2962fb460ec8faeb6dd5d4348fa5ddcf7e67cf11) regarding choice of worker:

> background_queue on prod uses prefork processes (three of them) and
> so using threads in the process should work nicely.
> 
> Using asyncio would not be faster than using threads, because we are
> not using an asyncio equivalent of the **requests** library to send
> requests, and so we would have to use threads to wrap requests anyway.


## Feature Flag

N/A. 

(This custom code is only executed for project spaces named `onse-iss`.)


## Product Description

N/A


## Safety Assurance

- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change
- [x] If QA is part of the safety story, the "Awaiting QA" label is used
- [x] I am certain that this PR will not introduce a regression for the reasons below


### Automated test coverage

Low-level utils are covered. High level functions are not.


### QA Plan

This code has been tested by calling it synchronously and asynchronously on Staging and locally. Staging test run has the same case load as Prod.


### Safety story

I have marked the risk as "Medium" and I am pinging @snopoke on this PR in case there is anything about running this with thread workers in the background queue on Prod that has implications that I have not thought of.


### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

kaapstorm committed:
Refactor: Move `for` loop one level up

kaapstorm committed:
Update cases in parallel using threads

background_queue on prod uses prefork processes (three of them) and
so using threads in the process should work nicely.

Using asyncio would not be faster than using threads, because we are
not using an asyncio equivalent of the **requests** library to send
requests, and so we would have to use threads to wrap requests anyway.

kaapstorm committed:
isort

snopoke commented on custom/onse/tasks.py:
```diff
@@ -38,8 +38,12 @@ def update_facility_cases_from_dhis2_data_elements(
+        with ThreadPoolExecutor(max_workers=MAX_THREAD_WORKERS) as executor:
+            futures = {executor.submit(set_case_updates, dhis2_server, cb)
+                       for cb in case_blocks}
+            for future in as_completed(futures):
```
can this not still be chunked?

snopoke reviewed (commented):
I have mixed feelings about this change. I think if it were in a management command using threads makes sense. Since this is already in a celery task I think it might be more appropriate to break this into multiple tasks:

```
case_ids = get_case_ids()
for chunk in chunked(case_ids):
    update_cases.delay(chunk)
```
(or use https://docs.celeryproject.org/en/stable/userguide/canvas.html#chunks)

Doing it this way means that the concurrency is limited by the celery queue configuration and . Currently there are 6 nodes running the 'background_queue' each with a concurrency of 3 which gives you max possible concurrency of 18 (depending on what else is also in the queue).

kaapstorm committed:
Process case blocks in chunks of 100

kaapstorm commented:
I really like the idea of using Celery chunks to break up the work.

kaapstorm commented:
The code is going to look cleaner using Celery chunks.

But I suspect we will get a smaller benefit, unless we're using Eventlet or gevent workers, because the background queue's prefork / process workers will still wait for DHIS2 to reply. And if we were going to use gevent workers, it isn't obvious to me which queue with gevent pooling we would use.

If using Python thread workers inside a Celery process worker feels clumsy (and I agree, it definitely feels like I'm mixing paradigms here), maybe I shouldn't do this. Most of the time it doesn't matter if this task takes hours to complete. It will happen late at night once a quarter.

kaapstorm committed:
Merge branch 'master' into nh/onse/by_case

kaapstorm committed:
Allow user to specify aggregation period

Useful if the data for the last quarter is still being collected.

kaapstorm committed:
Option to dump requests as curl commands

kaapstorm committed:
Retry 500 errors

kaapstorm committed:
Print notifications to STDERR

kaapstorm committed:
Fix is_500_error()

kaapstorm committed:
Useful script header

kaapstorm committed:
Revert option to dump requests as curl commands

kaapstorm committed:
Create class for storing case and building updates

TODO: Rename class

kaapstorm committed:
♼ Pull out success and error handling

kaapstorm committed:
♼ get_last_quarter() to previous_quarter()

kaapstorm committed:
If data not found, try previous quarters

kaapstorm committed:
Longer backoff, retry timeouts

kaapstorm commented:
I've added some commits to this branch to allow us to deal with errors from DHIS2, and connection timeouts.

The commits [Option to dump requests as curl commands](https://github.com/dimagi/commcare-hq/pull/28784/commits/f7b74c74039d8306056a2573b8d2afa8948f8b87) to [Revert option to dump requests as curl commands](https://github.com/dimagi/commcare-hq/pull/28784/commits/06b7feeae35292cac8ecf3947523a632c2b64f0e) are to reproduce errors and retry requests.

Once we could see the nature of DHIS2's data, we changed our approach. Instead of fetching data for a single quarter, we initially import data from the last five years for each case, and after that we just import data for the previous quarter.

The commit that finally resulted in consistently successful imports is the last commit, [Longer backoff, retry timeouts](https://github.com/dimagi/commcare-hq/pull/28784/commits/1a0e6e2de636132ddff7ccdbb2f6ca41f609a973).

You can find [the code that uses grouped Celery tasks here](https://github.com/dimagi/commcare-hq/blob/ac2c47864e609ee0d572bd49686e98805c0b3af9/custom/onse/tasks.py#L42-L47). But I would like to advocate for using threads after all:

Pros for using Celery:

* Code is simpler
* Maintains consistency by using Celery tasks within Celery tasks

Cons against using Celery:

* The choice of concurrency method -- processes, threads or green threads -- is managed with CommCare Cloud, not in the CommCare HQ codebase.

* This particular use case is light on processing, but benefits from concurrency because it spends a lot of time waiting for I/O. Using green threads or threads would optimise resources because they allow us to get the most out of a core. But throwing more processes at the problem means more cores are waiting for remote API responses; in other words, the benefit of concurrency comes at the cost of increasing inefficiency when we use processes.

* The choice of concurrency method is tightly constrained. To switch from using processes to using green threads on Prod would either mean creating a new queue just for the `set_case_updates()` task, or using the repeat_records queue, neither of which seem like the right approach.

* This task can take several hours to complete. Using the background queue for `update_facility_cases_from_dhis2_data_elements()` and the `set_case_updates()` tasks that it spawns can hold up several the queue workers for a noticeable amount of time. But if `update_facility_cases_from_dhis2_data_elements()` is only taking one Celery background queue worker, and the threads it spawns are running on the same core, that feels friendlier to other Celery background queue tasks.

Keen to hear your thoughts on this.

snopoke reviewed (commented):
The approach seems fine to me.

kaapstorm committed:
Improve iterating completed futures

kaapstorm committed:
Use relativedelta
